Add tutorial for in-place solve (#134)
JuliaSmoothOptimizers/JSOSuite.jl: docs/src/resolve.md

> # [Resolve and in-place solve](@id resolve)

Reviewer: ```suggestion
# [Re-solve and in-place solve](@id resolve)
```

It is very convenient to pre-allocate the memory used during the optimization of a given problem either from improved memory management or when re-solving the same or a similar problem.

Let us consider the following 2-dimensional unconstrained problem
```math
\begin{aligned}
\min \quad & f(x):= x₂ exp(x₁) 
\end{aligned}
```

> Using `JSOSuite` usual `minimize` function, the problem can be solved as follows

Reviewer: ```suggestion
Using `JSOSuite`’s `minimize` function, the problem can be solved as follows
```

```@example ex1
using JSOSuite

> f(x) = x[2]^2 * exp(x[1]^2)

Reviewer: This isn’t the same function as above.

x0 = ones(Float64, 2)
stats = minimize(f, x0)
```
Using L-BFGS, the problem is locally solved.

## In-place solve

> If we want to solve the same problem several time, for instance, for several initial guess it is recommended to use in-place solve.

Reviewer: ```suggestion
If we want to solve the same problem several times, for instance, for several initial guesses, it is recommended to use an in-place solve.
```

```@example ex1
using ADNLPModels, JSOSolvers, SolverCore
f(x) = x[2]^2 * exp(x[1]^2)
x0 = ones(Float64, 2)
nlp = ADNLPModel(f, x0) # or use JuMP

> solver = JSOSolvers.LBFGSSolver(nlp)

Reviewer: Maybe explain that you can’t just pass `f` anymore, but you need to create a model.

Author: I added a paragraph above to justify further. That would be something great to add too, the `solve!(solver, f, x0, stats; kwargs...)`.

stats = SolverCore.GenericExecutionStats(nlp)
solve!(solver, nlp, stats, x = x0)
```

> This deserves more explanation.

Reviewer: ```suggestion
This deserves more explanations.
```

The name of the solver structure and the corresponding package can be accessed via the DataFrame `JSOSuite.optimizers`.
```@example ex1
JSOSuite.optimizers[!, [:name_solver, :name_pkg]]
```
In our example, the solver LBFGS is implemented in `JSOSolvers.jl` and the solver structure is `LBFGSSolver`.

> Now, it is possible to reuse the memory allocated for the first solve for another round.

Reviewer: ```suggestion
Now, it is possible to reuse the memory allocated for the first solve for another round:
```

```@example ex1
# NLPModels.reset!(nlp) would also reset the evaluation counters of the model
reset!(solver)

> x02 = [1.4; 5.0] # another initial guess

Reviewer: ```suggestion
new_x0 = [1.4; 5.0] # another initial guess
```

> solve!(solver, nlp, stats, x = x02)

Reviewer: ```suggestion
solve!(solver, nlp, stats, x = new_x0)  # new solve with existing solver object
```

```

## In-place solve of a different problem

> It is also possible to reuse the allocated memory to solve another problem with the same characteristics.

Reviewer: ```suggestion
It is also possible to reuse the allocated memory to solve another problem with the same number of variables and constraints:
```
Question: must `nnzh` and `nnzj` also be the same?

Author: It depends on what the solver needs. If we pre-allocate some space for Jacobian and Hessian values, then yes.

```@example ex1
f2(x) = x[2]^2 + exp(x[1]^2)
x0 = ones(Float64, 2)
nlp = ADNLPModel(f2, x0) # or use JuMP
reset!(solver, nlp)
solve!(solver, nlp, stats)
```

## Allocation-free solvers

In order to measure, the amount of allocations made by the solvers the package NLPModelsTest.jl defines a set of test problems that are allocation-free.
```@example ex1
using NLPModelsTest, SolverCore, JSOSolvers
nlp = BROWNDEN(Float64)
solver = LBFGSSolver(nlp)
stats = GenericExecutionStats(nlp)
solve!(solver, nlp, stats)
@allocated solve!(solver, nlp, stats)
```

> Several of the pure Julia implementation available in JSOSuite have this property.

Reviewer: ```suggestion
Several of the pure Julia solvers available in JSOSuite have this property.
```

# [Re-solve and in-place solve](@id resolve)

It is very convenient to pre-allocate the memory used during the optimization of a given problem either for improved memory management or re-solving the same or a similar problem.

Let us consider the following 2-dimensional unconstrained problem
```math
\begin{aligned}
\min_x \quad & f(x):= x₂² exp(x₁²) 
\end{aligned}
```
Using `JSOSuite`’s `minimize` function, the problem can be solved as follows
```@example ex1
using JSOSuite
f(x) = x[2]^2 * exp(x[1]^2)
x0 = ones(Float64, 2)
stats = minimize(f, x0)
```

> Using L-BFGS, the problem is locally solved.

Author: I found a way to show the console output :), which justifies it.
